Use a single construct block

rainwoodman opened the pull request:
In a construct block we know the construct setter of attributes have all been called. Thus we initialize the immutable property values.

rainwoodman committed:
Use a single construct block.

This actually works. Notice how test gobject_construct is fixed.

reshape shall return an array on the same data. it shall not
reallocate a new data.

rainwoodman committed:
add comments.

rainwoodman committed:
assert dimension less than 32.

arteymix commented on src/array.vala:
```diff
@@ -2,70 +2,84 @@ using GLib;
+
+        /* We will copy the in-values from g_object_new */
+        for (var i = 0; i < _dimension; i ++) {
+            _shape[i] = _shape_in[i];
```
```vala
Memory.copy (_shape, _shape_in, size_of (size_t) * _dimension);
```

arteymix commented on src/array.vala:
```diff
@@ -2,70 +2,84 @@ using GLib;
+
+        /* the input pointers from gobject is no longer useful, void them.*/
+        _shape_in = null;
+        _strides_in = null;
```
I love it!

arteymix commented on src/array.vala:
```diff
@@ -2,70 +2,84 @@ using GLib;
 
 public class Vast.Array : Object
 {
+    /* GType of the scalar elements, immutable */
```
You should use multi-line comments with `/**` prefix, otherwise Valadoc won't like it.

arteymix commented on src/array.vala:
```diff
@@ -2,70 +2,84 @@ using GLib;
-        }
-    }
+    /* GObject that owns the memory buffer for storage of scalar elements */
+    public Bytes? data {get; construct; }
```
Could we have a `default = null` here?

arteymix commented:
Maybe we should wait for adding comments until we commit to something stable. It was a real pain in Valum to update them afterward.

Otherwise everything looks good!

rainwoodman committed:
promote origin to a pointer. cache it in Iterator.

This way origin can be any memory location. data doesn't need to be
a GBytes anymore.

arteymix commented on src/array.vala:
```diff
@@ -41,11 +41,8 @@ public class Vast.Array : Object
-    /* pointer to the begining memory buffer, */
-    internal unowned uint8* _cached_data;
+    /* pointer to the memory location of 0th element */
+    public uint8 * origin {get; construct; default = null;}
```
What I don't like here is that we publicly expose a pointer. This won't work if the array is backed by, let's say, a network or a file storage. Couldn't we do the computed property trick around `get_data ()` pointer and have an internal base pointer?

arteymix replied:
For the property, we could have a public `size_t` and an internal computed pointer from get and construct without changing any name.

arteymix replied:
If we want it to work for `Iterator`, the `_origin` property would have to be internal. In C we could probably hijack the private struct instead of exposing a public field ;)

arteymix replied:
That, or recomputing it from `array.data.get_data () - array.origin`.

arteymix replied:
It seems to me that the day we'll extract an abstract class, we'll have to have specialized `Iterator` implementation as well to take advantage of the specific bits of each implementation.

arteymix commented on src/iterator.vala:
```diff
@@ -19,6 +19,12 @@ public class Vast.Iterator : Object
         get { return _offset; }
     }
 
+    public uint8 * origin {get; private set;}
```
I would not use a property here because it's just forwarding the array's origin. It's okay to cache the origin pointer for faster access though.

arteymix commented on src/array.vala:
```diff
@@ -2,70 +2,82 @@ using GLib;
+            _size *= _shape[i];
+        }
+
+        _data = _data ?? new Bytes (new uint8[scalar_size * _size]);
```
This is going to break the current hack for PyGObject (a.k.a empty array). Maybe we could wait to get a proper fix for introspection and keep allocating in the `Array` constructor?

I also agree that once constructed, the data should be fully allocated.

arteymix commented on src/array.vala:
```diff
@@ -84,7 +96,8 @@ public class Vast.Array : Object
                   ssize_t[] strides = {},
                   Bytes?    data    = null,
-                  size_t    origin  = 0)
+                  uint8*    origin = null)
```
To add to my point from below, it's impossible to define a proper origin if you let the array allocate the bytes. It should definately by `size_t` from the public eye.

arteymix committed:
array: Expose 'origin' as 'size_t', but back it with a pointer

From a public perspective, passing a pointer does not always make sense.
In the case of a 'null' data, there would be no way to know in advance
where it would be allocated.

Instead pass a 'size_t' and compute an internal '_baseptr'.

In 'Iterator', just recompute '_baseptr' from the array's data and
origin.

arteymix committed:
array: Copy '_shape_in' using 'Memory.copy'

arteymix commented:
I'm okay with all this and breaking PyGObject is fine since we'll actively look for a fix.

arteymix committed:
array: Add explicit default values for 'scalar_type' and 'bytes'

arteymix commented on src/array.vala:
```diff
@@ -2,70 +2,86 @@ using GLib;
 {
-    public Type scalar_type { get; construct; }
+    /* GType of the scalar elements, immutable */
+    public Type scalar_type { get; construct; /* implicitly typeof (void) */ }
```
This is a really tricky thing because when I set it explicitly, it crashes. It would probably be a good idea to investigate.

arteymix commented on tests/vast-test.vala:
```diff
@@ -60,14 +58,13 @@ int main (string[] args) {
 
-        var b = a.reshape ({2, 2, 2, 4});
+        size_t [] shape = {2, 2, 2, 4};
+        var b = Object.new (typeof (Vast.Array), "dimension", shape.length, "shape", shape) as Vast.Array;
```
@rainwoodman By the way, here, you can use `dimension: shape.length, shape: shape` notation. It's supported by Vala for alternating varidic arguments.

arteymix commented on tests/vast-test.vala:
```diff
@@ -60,14 +58,13 @@ int main (string[] args) {
-        assert (null == a.data);
-        assert ("dtype: void, dsize: %lu, dimension: 0, shape: (), strides: (), mem: 0B".printf (sizeof (void)) == a.to_string ());
+        assert (null != a.data);
+        //assert ("dtype: void, dsize: %lu, dimension: 0, shape: (), strides: (), mem: 0B".printf (sizeof (void)) == a.to_string ());
```
We should probably update this assertion instead.

arteymix committed:
array: Remove check for '_data == null' in 'reshape'

Update the PyGObject test case. We need to fix 'shape' and 'strides' at
construct-time, not rely on a hack for shaping arrays.

arteymix committed:
array: Nullify '_shape_in' and '_strides_in' on-need

Ensure that we are really dealing with a scalar-like array if a 'null'
shape is given.

arteymix commented:
Hmm, we could basically avoid summing `_baseptr + _offset` in `Iterator` and operate directly on `_baseptr`. Still, we keep `offset` for the public eye.

rainwoodman committed:
update the comments.

rainwoodman committed:
update assertion in test.

arteymix commented:
You're right that there's a lot of `_`. Maybe we could avoid the getter overhead strictly for critical regions?

Also, it might just be optimized away, although I'm not exactly sure if it's done for public symbols.

rainwoodman committed:
one more comment.

rainwoodman commented:
I think it is ok for accessing the variables -- because if we define an attribute '_' is automatically defined by vala. But the '_' prefixed functions are what concerns me. I think we should be able to move the xxx_from_xxx functions for slicing into a single struct. I'll merge this one and deal with that in a new pr.
